enginetest/queries/json_scripts.go

```diff
@@ -17,11 +17,112 @@ package queries
 import (
 	querypb "github.com/dolthub/vitess/go/vt/proto/query"
 
+	"github.com/dolthub/go-mysql-server/sql/expression/function/json"
+
 	"github.com/dolthub/go-mysql-server/sql"
 	"github.com/dolthub/go-mysql-server/sql/types"
 )
 
 var JsonScripts = []ScriptTest{
+	{
+		Name: "json_type scripts",
+		Assertions: []ScriptTestAssertion{
+			{
+				Query: "select JSON_TYPE(CAST(1 AS JSON))",
+				Expected: []sql.Row{
+					{"INTEGER"},
+				},
+			},
+			{
+				Query: `select JSON_TYPE(CAST("1" AS JSON))`,
+				Expected: []sql.Row{
+					{"INTEGER"},
+				},
+				// Because standard JSON doesn't have separate int and float types,
+				// golang's JSON parser parses the "1" as a float.
+				SkipResultsCheck: true,
+			},
+			{
+				Query: `select JSON_TYPE(CAST("\"1\"" AS JSON))`,
+				Expected: []sql.Row{
+					{"STRING"},
+				},
+			},
+			// Casting without quotes: `321.4` is parsed as a decimal, then wrapped in a JSON Document.
+			{
+				Query: "select JSON_TYPE(CAST(321.4 AS JSON))",
+				Expected: []sql.Row{
+					{"DECIMAL"},
+				},
+			},
+			// Casting with quotes: The string value is parsed as JSON, resulting in a wrapped double.
+			{
+				Query: `select JSON_TYPE(CAST("321.4" AS JSON))`,
+				Expected: []sql.Row{
+					{"DOUBLE"},
+				},
+			},
+			{
+				Query: `select JSON_TYPE(CAST("\"321.4\"" AS JSON))`,
```

Comment:
we've had weird type edge cases with exponentials before, ex `1e2`, `"1e2"`, `1.0e2`, ...

Reply:
Added additional tests.